pkg/sql/opt/metadata.go

```diff
@@ -726,8 +726,9 @@ func (md *Metadata) DuplicateTable(
 		partialIndexPredicates:        partialIndexPredicates,
 		indexPartitionLocalities:      tabMeta.indexPartitionLocalities,
 		checkConstraintsStats:         checkConstraintsStats,
-		notVisibleIndexMap:            tabMeta.notVisibleIndexMap,
 	}
+	newTabMeta.indexVisibility.cached = tabMeta.indexVisibility.cached
+	newTabMeta.indexVisibility.notVisible = tabMeta.indexVisibility.notVisible
```

Comment:
This is an interesting case that I didn't think about during the initial review. When we duplicate a table in the memo, do we want the same indexes to be visible or does it not matter?

Initially, I'm leaning towards having the same visibility. But unfortunately, this isn't guaranteed to achieve that because the "small" regime of these sets are passed by value, so if `IsIndexNotVisible` has not been called for an index before the table is duplicated, the two tables may generate and cache different invisibility booleans.

Do you think this would be a problem? Maybe it's not sufficient enough to worry about. I'd be worried about rules that duplicate tables specifically to use an index. What happens if that index is then not visible? Is that even possible, or would `IsIndexNotVisible` have been called before the table was duplicated?

Reply:
That's a good point... I'll benchmark the impact of making those pointers instead, since that seems like the safest option.

Reply:
https://github.com/cockroachdb/cockroach/pull/112727